pkg/tbtc/maintainer/bitcoin_difficulty.go

```diff
@@ -2,54 +2,271 @@ package maintainer
-			logger.Info("Bitcoin difficulty maintainer is working")
 		}
+	}
+}
 
-		time.Sleep(1 * time.Second)
+// proveEpochs proves Bitcoin blockchain epochs in the Bitcoin difficulty chain.
+func (bdm *BitcoinDifficultyMaintainer) proveEpochs(ctx context.Context) error {
+	if err := bdm.verifySubmissionEligibility(); err != nil {
+		return fmt.Errorf(
+			"cannot proceed with proving Bitcoin blockchain epochs [%v]",
+			err,
+		)
 	}
+
+	for {
+		if err := bdm.proveNextEpoch(); err != nil {
+			return fmt.Errorf("cannot prove Bitcoin blockchain epoch [%v]", err)
+		}
+
+		select {
+		case <-time.After(bdm.epochProvenBackOffTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// verifySubmissionEligibility verifies whether a maintainer is eligible to
+// submit block headers to the Bitcoin difficulty chain.
+func (bdm *BitcoinDifficultyMaintainer) verifySubmissionEligibility() error {
+	isReady, err := bdm.chain.Ready()
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether genesis has been performed [%v]",
+			err,
+		)
+	}
+
+	if !isReady {
+		return fmt.Errorf(
+			"genesis has not been performed in the Bitcoin difficulty chain",
+		)
+	}
+
+	authorizationRequired, err := bdm.chain.AuthorizationRequired()
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether authorization is required to submit "+
+				"block headers [%v]",
+			err,
+		)
+	}
+
+	if !authorizationRequired {
+		return nil
+	}
+
+	maintainerAddress := bdm.chain.Signing().Address()
+
+	isAuthorized, err := bdm.chain.IsAuthorized(maintainerAddress)
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether Bitcoin difficulty maintainer is authorized "+
+				"to submit block headers [%v]",
+			err,
+		)
+	}
+
+	if !isAuthorized {
+		return fmt.Errorf(
+			"bitcoin difficulty maintainer has not been authorized to submit " +
+				"block headers",
+		)
+	}
+
+	return nil
+}
+
+// proveNextEpoch proves a single Bitcoin blockchain epoch in the Bitcoin
+// difficulty chain if there is a Bitcoin blockchain epoch to be proven.
+func (bdm *BitcoinDifficultyMaintainer) proveNextEpoch() error {
+	// The height of the Bitcoin blockchain.
+	currentBlockNumber, err := bdm.btcChain.GetCurrentBlockNumber()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current block number [%v]",
+			err,
+		)
+	}
+
+	// The current epoch proven in the Bitcoin difficulty chain.
+	currentEpoch, err := bdm.chain.CurrentEpoch()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current epoch [%v]",
+			err,
+		)
+	}
+
+	// The number of blocks required for each side of a retarget proof.
+	proofLength, err := bdm.chain.ProofLength()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get proof length [%v]",
+			err,
+		)
+	}
+
+	// The new epoch to be proven in the Bitcoin difficulty chain.
+	newEpoch := currentEpoch + 1
+
+	// Height of the first block of the new epoch.
+	newEpochHeight := newEpoch * bitcoinEpochLength
+
```

Comment:
I think this assertion is unnecessary, because the smart contract guarantees  `0` < `proofLength` < `2016` and we know `newEpochHeight` will always be `2016` or higher, because it's obtained like that:
```
newEpoch := currentEpoch + 1
newEpochHeight := newEpoch * bitcoinDifficultyEpochLength
```
and `currentEpoch` is always 0 or more.